Added SPLIT_V op to micro

yair-ehrenwald opened the pull request:
Added split_v to micro, supports cases where the output tensors were pre-allocated in the model.
Ported tests from TFLite

Commit:
Added prototype

yair-ehrenwald committed:
Added prototype

yair-ehrenwald committed:
Added TANH op and test (added to elementwise testing)

Commit:
Added TANH op and test (added to elementwise testing)

Commit:
Made float syntax consistent

yair-ehrenwald committed:
Made float syntax consistent

Commit:
made float syntax consistent

Commit:
made float syntax consistent

yair-ehrenwald committed:
made float syntax consistent

Commit:
made float syntax consistent

yair-ehrenwald committed:
made float syntax consistent

yair-ehrenwald committed:
made float syntax consistent

yair-ehrenwald committed:
fixed float syntax inconsistency

Commit:
fixed float syntax inconsistency

yair-ehrenwald committed:
fixed float syntax inconsistency

yair-ehrenwald committed:
Updated floats

yair-ehrenwald committed:
Updated floats

yair-ehrenwald committed:
Updated floats

yair-ehrenwald committed:
Merge branch 'master' of https://github.com/yair-ehrenwald/tensorflow

yair-ehrenwald committed:
Merge branch 'master' of https://github.com/yair-ehrenwald/tensorflow

yair-ehrenwald committed:
Merge branch 'master' of https://github.com/yair-ehrenwald/tensorflow

yair-ehrenwald committed:
fixed typo

yair-ehrenwald committed:
Add split_v to micro

* started working on split_v

* Added split_v support and tests for micro

yair-ehrenwald committed:
Revert "Add split_v to micro" (#2)

This reverts commit ac28ce209668057d9b6f6b2a139ca5ab28476e78.

yair-ehrenwald committed:
Add files via upload

Add split_v.cc and split_v_test.cc

yair-ehrenwald committed:
Added SPLIT_V prototype to micro_ops.h

yair-ehrenwald committed:
Added SPLIT_V

petewarden commented on tensorflow/lite/micro/kernels/split_v.cc:
```diff
@@ -0,0 +1,257 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#include <vector>
```
Use of std::vector is usually a red flag on Micro, since it relies on dynamic memory allocation. This does make porting kernels like this tough, but we are open to approaches like capping the maximum number of tensors, like we do with concatenation: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/kernels/concatenation.cc#L31

We do also now have the scratch memory API available if temporary buffers are required, but we try to minimize the memory usage generally.

yair-ehrenwald replied:
I removed dependencies for std:vector. It's used in ResizeOutputTensors() which we don't need here. The output tensors have already been created the right size directly from the model.

yair-ehrenwald replied:
I removed all usage of dynamic memory allocation, followed the same flow as in the regular split  operation. No std::vector or any other  dynamic allocation necessary - the output tensors are already allocated in the model.

petewarden replied:
Great! Can you also remove the #include <vector> since it's no longer needed? Thanks.

yair-ehrenwald replied:
Done :)

petewarden commented on tensorflow/lite/micro/kernels/split_v_test.cc:
```diff
@@ -0,0 +1,656 @@
+#include "tensorflow/lite/micro/testing/micro_test.h"
+#include "tensorflow/lite/micro/testing/test_utils.h"
+
+namespace tflite {
```
The whitespace and indentation on this file seem to be unusual. Could you check what's going on, and possibly use clang-format to standardize it?

yair-ehrenwald replied:
Done.

yair-ehrenwald committed:
Removed calls to ResizeOutputTensors() from split_v and fixed whitespaces in split_v_test

Calls to ResizeOutputTensors() aren't required in micro since the output buffers have already been allocated from the model

yair-ehrenwald committed:
Removed dynamic memory usage and stl types from split_v

yair-ehrenwald committed:
Removed unnecessary #include

rthadur commented on tensorflow/lite/micro/kernels/split_v.cc:
```diff
@@ -0,0 +1,168 @@
+  // if not const return error - unsupported
+  if (!IsConstantTensor(op_context.axis) ||
+      !IsConstantTensor(op_context.size_splits)) {
+    context->ReportError(
```
`TF_LITE_KERNEL_LOG should be used instead of context->ReportError, so log strings can be removed to save space if needed.`

yair-ehrenwald replied:
@rthadur  Done.

rthadur commented on tensorflow/lite/micro/kernels/split_v.cc:
```diff
@@ -0,0 +1,168 @@
+      break;
+    }
+    default:
+      context->ReportError(context, "Type %s currently not supported.",
```
`TF_LITE_KERNEL_LOG should be used instead of context->ReportError, so log strings can be removed to save space if needed.`

yair-ehrenwald replied:
Done

gbaned commented:
@yair-ehrenwald  Can you please check @rthadur's comments and keep us posted. Thanks!

yair-ehrenwald committed:
Changed error to use TF_LITE_KERNEL_LOG

yair-ehrenwald committed:
Fixed error output to use TF_LITE_KERNEL_LOG

gbaned commented:
@yair-ehrenwald Can you please resolve conflicts? Thanks!

yair-ehrenwald commented:
@gbaned Unless I'm missing something the only conflict is in
tensorflow/lite/micro/kernels/all_ops_resolver.cc
which has the addition of the split_v registration?

gbaned commented:
@petewarden Can you please assist on above comments from @yair-ehrenwald. Thanks!

rthadur commented:
@yair-ehrenwald To proceed, you should just sync your master branch to tensorflow HEAD, rebase your change on top of that, then push your change branch back.

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

gbaned commented:
@yair-ehrenwald, Any update on this PR? Please. Thanks!

yair-ehrenwald commented:
@gbaned  Should be done shortly, thanks :)

yair-ehrenwald committed:
Merge remote-tracking branch 'upstream/master'

yair-ehrenwald committed:
Merge branch 'master' of https://github.com/yair-ehrenwald/tensorflow

gbaned commented:
@yair-ehrenwald  Can you please sign CLA. Thanks!

yair-ehrenwald commented:
@googlebot I fixed it

yair-ehrenwald commented:
@gbaned  Should be done now, with no conflicts :)

gbaned commented:
@yair-ehrenwald Can you please sign CLA. Thanks!

yair-ehrenwald commented:
> @yair-ehrenwald Can you please sign CLA. Thanks!

@gbaned Hi, I signed it, should be ok.

yair-ehrenwald commented:
@googlebot I fixed it.

gbaned commented:
@yair-ehrenwald  It still shows CLA is pending , can you please sign CLA. Thanks!

OSMaxwell commented:
Could someone give some insight on replacing <vector> and dynamic memory allocation in Micro implementation?

gbaned commented:
@yair-ehrenwald It still shows CLA is pending , can you please sign CLA. Thanks!

gbaned commented:
@yair-ehrenwald, Any update on this PR? Please. Thanks!

yair-ehrenwald commented:
> @yair-ehrenwald, Any update on this PR? Please. Thanks!

@gbaned  I think I understand why it says no CLA, I will fix the issue and open a new PR (and link to this one) shortly.

gbaned commented:
@yair-ehrenwald, Any update on this PR? Please. Thanks!

yair-ehrenwald commented:
https://github.com/tensorflow/tensorflow/pull/42452
